### ISSUE TITLE
PackageCompiler will not compile my (simple) Julia program to a static binary due to ArgParse errors

### ISSUE THREAD

**Issue by isene** [JuliaLang/PackageCompiler.jl#76]
I am trying to compile my program to an executable on Linux using PackageCompiler - but can't get it to work on either Julia 0.6 or 0.7. In both cases it's ArgParse that creates the error, but different errors for the two versions. See the attached files for detail (I renamed my program to be compiled here from O6.jl to O6.txt due to limitations of uploading .jl files)

[O6.txt](https://github.com/JuliaLang/PackageCompiler.jl/files/2088254/O6.txt)
[julia06.txt](https://github.com/JuliaLang/PackageCompiler.jl/files/2088249/julia06.txt)
[julia07.txt](https://github.com/JuliaLang/PackageCompiler.jl/files/2088250/julia07.txt)

**Comment by NHDaly** [comment]
Ah, this one is easy! The problem isn't related to ArgParse at all, your julia program is simply missing `julia_main`, which is required in order to be compiled. Adding this to the bottom of your file `O6.txt` is enough to fix it:
```
Base.@ccallable function julia_main()::Cint
  O6()
  return 0
end
```

**Comment by NHDaly** [comment]
Ah, this one is easy! The problem isn't related to ArgParse at all, your julia program is simply missing `julia_main`, which is required in order to be compiled. Adding this to the bottom of your file `O6.txt` is enough to fix it:

```
Base.@ccallable function julia_main()::Cint
  O6()
  return 0
end
```

I've just noticed that's not mentioned anywhere on the README at all... I will send a PR up to fix that now.

EDIT: oops, sorry, I hit "shift-enter" which apparently submits the comment, so it was submitted before I finished typing. Sorry!

**Comment by NHDaly** [comment]
(You can see the problem here, in the output from `juliac`):

```
...
/tmp/ccVlecLr.o: In function `main':
program.c:(.text+0x118): undefined reference to `julia_main'
collect2: error: ld returned 1 exit status
ERROR: LoadError: failed process: Process(`cc '-DJULIAC_PROGRAM_LIBNAME="O6.so"' -o /home/geir/Dropbox/G/Tech/Julia/builddir/O6 /home/geir/.julia/v0.6/PackageCompiler/examples/program.c /home/geir/Dropbox/G/Tech/Julia/builddir/O6.so -m64 -std=gnu99 -I/home/geir/Dropbox/Dl/tmp/julia-06/include/julia -DJULIA_ENABLE_THREADING=1 -fPIC -L/home/geir/Dropbox/Dl/tmp/julia-06/lib -Wl,--export-dynamic -Wl,-rpath,/home/geir/Dropbox/Dl/tmp/julia-06/lib -Wl,-rpath,/home/geir/Dropbox/Dl/tmp/julia-06/lib/julia -ljulia '-Wl,-rpath,$ORIGIN'`, ProcessExited(1)) [1]
...
```

But there's a lot of output, so that message is hard to find.

**Comment by lucatrv** [comment]
I assumed there was an issue with `ArgParse` because I could not statically compile `juliac.jl` (but had never really bothered to better analyze this issue before...) now I found out that the issue is actually due to `using PackageCompiler` and to calling `julia_main(ARGS)` within `juliac.jl`.
It appears to me that when Julia statically compiles a script with the command `$julia_cmd --output-o $o_file -e $expr`, it also executes it. This can easily be verified by compiling a simple script with just one line `println("test")`: "test" is shown while compiling.
Now I think this is why `juliac.jl` cannot be statically compiled, and maybe it is a Julia bug, do you agree?

**Comment by NHDaly** [comment]
Hmm, no I don't think so! I think it _needs_ to execute the script, otherwise it might not know which code to compile!

Take the following trivial example:

```julia
if VERSION >= v"0.7"
  include("mycode-0.7")
else
  include("mycode-0.6")
end
```

----

I have been thinking about global, "script-like" code as basically static, compile-time only logic, similar to c++ preprocessor macros, but way better because it's just normal, first class julia code.

**Comment by isene** [comment]
Thanks. Adding that julia_main at the end there got me an executable. And it does run and gives the correct result. But it also gives this warning every time I run it (how do I remove this?):

WARNING: redefining constant JULIA_HOME

Also, being new to Julia, this looks ugly: julia_main(ARGS::Vector{String})::Cint
Why is all that needed?

**Comment by lucatrv** [comment]
@NHDaly one option could be to be able to detect programmatically if a script is being statically compiled, and based on that either do or do not execute parts of code.
At present I am not able either to check if the script is being statically compiled, or if it has been statically compiled,  `get(ENV, "COMPILE_STATIC", "false") == "false"` does not seem to work (see `juliac.jl`).

**Comment by NHDaly** [comment]
@lucatrv Yeah I saw that as well. Actually, you can fix that by calling `build_executable` through `withenv` to set `"COMPILE_STATIC"` to `true`! I don't know how that code got there in the first place though..?

I've opened an issue specific to what you're trying to do here: https://github.com/JuliaLang/PackageCompiler.jl/issues/78 Let's talk more about it there! :)

**Comment by NHDaly** [comment]
@isene: Also, yay! **welcome to julia!!** :)

![giphy](https://user-images.githubusercontent.com/1582097/41418679-39e20724-6fbe-11e8-94b8-6323d8c44875.gif)

**Comment by isene** [comment]
Amazing comment. Thanks for taking the time to explain this so thoroughly.

**Comment by lucatrv** [comment]
Can this issue be closed?

**Comment by blaiseli** [comment]
@lucatrv May I reactivate this not-yet-closed issue? The title corresponds to my initial issue.

I had a program containing a `julia_main` function, but this function was then called, which caused errors with `ArgParse` in the message.

Thanks to the above explanations, I understand now that my program should just define `julia_main`, but not execute it, and I imagine that execution will be handled by some C code somewhere.

So I commented out the call to `julia_main`.

Now, with `julia_main` still being defined, but not called, running `build_executable` on my program surprisingly results in the following error:

```
Julia program file:
  "/home/bli/src/qaf_demux/Julia/QafDemux/bin/qaf_demux.jl"
C program file:
  "/home/bli/.julia/packages/PackageCompiler/CJQcs/examples/program.c"
Build directory:
  "/home/bli/src/qaf_demux/Julia/QafDemux/builddir"
/tmp/ccAAtpnB.o: In function `main':
program.c:(.text+0x153): undefined reference to `julia_main'
collect2: error: ld returned 1 exit status
ERROR: LoadError: failed process: Process(`cc '-DJULIAC_PROGRAM_LIBNAME="qaf_demux.so"' -o qaf_demux /home/bli/.julia/packages/PackageCompiler/CJQcs/examples/program.c qaf_demux.so -std=gnu99 -I/home/bli/src/julia/usr/include/julia -DJULIA_ENABLE_THREADING=1 -fPIC -L/home/bli/src/julia/usr/lib -Wl,--export-dynamic -Wl,-rpath,/home/bli/src/julia/usr/lib -Wl,-rpath,/home/bli/src/julia/usr/lib/julia -ljulia -m64 '-Wl,-rpath,$ORIGIN'`, ProcessExited(1)) [1]

Stacktrace:
```

etc.

I double-checked that `julia_main` is defined in my code:

```
$ grep "julia_main" /home/bli/src/qaf_demux/Julia/QafDemux/bin/qaf_demux.jl
function julia_main()::Cint
# julia_main()
```

Here is my build script:

```
$ cat deps/build.jl
import Pkg
Pkg.add("ArgParse")
Pkg.add("IterTools")
Pkg.add("FASTX")
Pkg.add("CodecZlib")
Pkg.add("PackageCompiler")
using PackageCompiler
build_executable(joinpath(@__DIR__, "../bin/qaf_demux.jl"), "qaf_demux")
```

(By the way, is that a correct way to write a build script? I couldn't find documentation giving examples of what kind of things should be put in a build script, just documentation saying that a project can have one in `deps/build.jl`, and that it will be executed upon `Pkg.build()`.)

And I execute it as follows:
```
$ julia deps/build.jl
```

All the above is done from the `QafDemux` directory, and I'm running julia 1.2:

```
$ julia --version
julia version 1.2.0
```
